components/script/xpath/eval_value.rs

```diff
@@ -0,0 +1,201 @@
+            Value::String(ref s) => str_to_num(s),
+            Value::Nodeset(..) => str_to_num(&self.string()),
+        }
+    }
+
+    pub fn into_number(self) -> f64 {
+        self.number()
+    }
+
+    pub fn string(&self) -> string::String {
+        match *self {
+            Value::Boolean(v) => v.to_string(),
+            Value::Number(n) => {
+                if n.is_infinite() {
+                    if n.signum() < 0.0 {
+                        "-Infinity".to_owned()
+                    } else {
+                        "Infinity".to_owned()
+                    }
+                } else if n == 0.0 {
+                    // catches -0.0 also
+                    0.0.to_string()
+                } else {
+                    n.to_string()
+                }
+            },
+            Value::String(ref val) => val.clone(),
+            Value::Nodeset(ref nodes) => match nodes.document_order_first() {
+                Some(n) => n.GetTextContent().unwrap_or_default().to_string(),
+                None => "".to_owned(),
+            },
+        }
+    }
+
+    pub fn into_string(self) -> string::String {
+        match self {
+            Value::String(val) => val,
+            other => other.string(),
+        }
+    }
+}
+
+macro_rules! from_impl {
+    ($raw:ty, $variant:expr) => {
+        impl From<$raw> for Value {
+            fn from(other: $raw) -> Value {
+                $variant(other)
+            }
+        }
+    };
+}
+
+from_impl!(bool, Value::Boolean);
+from_impl!(f64, Value::Number);
+from_impl!(String, Value::String);
+impl<'a> From<&'a str> for Value {
+    fn from(other: &'a str) -> Value {
+        Value::String(other.into())
+    }
+}
+from_impl!(Vec<DomRoot<Node>>, Value::Nodeset);
+
+macro_rules! partial_eq_impl {
+    ($raw:ty, $variant:pat => $b:expr) => {
+        impl PartialEq<$raw> for Value {
+            fn eq(&self, other: &$raw) -> bool {
+                match *self {
+                    $variant => $b == other,
+                    _ => false,
+                }
+            }
+        }
+
+        impl PartialEq<Value> for $raw {
+            fn eq(&self, other: &Value) -> bool {
+                match *other {
+                    $variant => $b == self,
+                    _ => false,
+                }
+            }
+        }
+    };
+}
+
+partial_eq_impl!(bool, Value::Boolean(ref v) => v);
+partial_eq_impl!(f64, Value::Number(ref v) => v);
+partial_eq_impl!(String, Value::String(ref v) => v);
+partial_eq_impl!(&str, Value::String(ref v) => v);
+partial_eq_impl!(Vec<DomRoot<Node>>, Value::Nodeset(ref v) => v);
+
+pub trait NodesetHelpers {
+    /// Returns the node that occurs first in [document order]
+    ///
+    /// [document order]: https://www.w3.org/TR/xpath/#dt-document-order
+    fn document_order_first(&self) -> Option<DomRoot<Node>>;
+    fn document_order(&self) -> Vec<DomRoot<Node>>;
+    fn document_order_unique(&self) -> Vec<DomRoot<Node>>;
+}
```

Comment:
I think there's an existing helper for iterating over a tree in document order, though I can't seem to find it right now...

Reply:
There `Node::traverse_preorder()` but note that here we don't have a tree of Nodes, it's a list of Nodes that can be from anywhere in the tree.